src/gwt/acesupport/loader.js

```diff
@@ -253,33 +254,192 @@ oop.inherits(RStudioUndoManager, UndoManager);
+};
+oop.inherits(RStudioFontMetrics, FontMetrics);
+
+(function() {
+
+   this.checkForSizeChanges = function(size) {
+      // compute size if we weren't given the size
+      if (typeof(size) === "undefined") {
+         size = this.$measureSizes();
+      }
+
+      // don't attempt to recompute when container is hidden
+      if (size && (size.height === 0 || size.width === 0)) {
+         return;
+      }
+
+      // update cached size and perform check for size
+      this.$cachedSize = size;
+      FontMetrics.prototype.checkForSizeChanges.call(this, size);
+   };
+
+}).call(RStudioFontMetrics.prototype);
 
 // RStudioRenderer ----
 
-var RStudioRenderer = function(container, theme) {
+// queue of virtual renderers that need to be resized (redrawn)
+var ResizeQueue = [];
+
+// method called on a timer to redraw all renderers that need it
+var ProcessResizeQueue = function() {
+   // sort the resize queue so that elements closer to the top get priority. if
+   // we ever need to support opening a document with the viewport positioned
+   // somewhere other than the top by default, this would need to take
+   // scrollTop and friends into consideration
+   ResizeQueue.sort(function(a, b) {
+      return a.pos - b.pos;
+   });
+
+   // walk the resize queue and schedule the actual work of resizing each
+   // element; we do this on a zero timer so that it doesn't block the UI if
+   // e.g. 200 resizes need to get processed at once
+   for (var idx = 0; idx < ResizeQueue.length; idx++) {
+      window.setTimeout(function(renderer, args) {
+         // call base class resize method
+         Renderer.prototype.onResize.apply(renderer, args);
+      }, 0, ResizeQueue[idx].renderer, ResizeQueue[idx].args);
+   }
+
+   // clear queue
+   ResizeQueue = [];
+   ResizeQueueTimer = 0;
+};
+
+// id of window timer to process resize queue
+var ResizeQueueTimer = 0;
+
+var RStudioRenderer = function(container, theme, fontMetrics) {
    Renderer.call(this, container, theme);
+
+   // use caller-supplied font metrics when specified
+   if (fontMetrics) {
+      var self = this;
+
+      // destroy original font metrics object (so we don't have a dangling observer)
+      if (self.$fontMetrics) {
+         self.$fontMetrics.destroy();
+      }
+
+      // replace with caller-supplied object
+      self.$fontMetrics = fontMetrics;
+      self.$textLayer.$setFontMetrics(fontMetrics);
+
+      // typically the font metrics change is what triggers a redraw when the
+      // size becomes nonzero; this is needed so that e.g. an editor that was
+      // initially rendered in an unselected tab can re-render itself with the
+      // correct metrics when it becomes visible.
+      //
+      // without font the font metrics system, we need another way to trigger
+      // that redraw, so we implement a direct resize observer here.
+      self.$cachedHeight = 0;
+      self.$resizeObserver = new window.ResizeObserver(function(e) {
```

Comment:
Are these APIs available on IE11? (Do we need to worry about IE11 for visual mode, or will we drop support for that in time for the 1.4 release?)